### ISSUE TITLE
Sequencer - Gradient check failing on variable-length sequences

### ISSUE THREAD

**Issue by mpienkosz** [Element-Research/rnn#204]
I'm trying to train RNN for text classification-like task. However, I'm having a problem while using Sequencer-based model and variable-length sequences - gradient check is failing. Here is the sample code to illustrate the problem:

```lua
require 'rnn'
require 'optim'

-- Hyperparameters
local inputSize = 5
local hiddenSize = 5
local nClasses = 4
local nIndex = 10
local maxSeqLen = 20
local nSamples = 50
local nEpochs = 10

-- Creating dummy dataset
local sentences = {}
local targets = {}
for i=1,nSamples do
  local seqLen = torch.random(maxSeqLen)
  local seq = torch.Tensor(seqLen):random(nIndex)
  local target = torch.random(nClasses)
  sentences[i] = seq
  targets[i] = target
end

-- Defining model
local lookupTable = nn.LookupTable(nIndex, inputSize)
local rnn = nn.Sequential()
  :add(nn.SplitTable(1,2))
  :add(nn.Sequencer(nn.FastLSTM(inputSize, hiddenSize)))
  :add(nn.SelectTable(-1))
  :add(nn.Linear(hiddenSize, nClasses))
  :add(nn.LogSoftMax())  
local criterion = nn.ClassNLLCriterion()
local params, gradParams = rnn:getParameters()

-- Training and gradient checking
for e=1,nEpochs do
  for i=1,nSamples do
    xlua.progress(i, nSamples)
    local feval = function(x)
      rnn:zeroGradParameters()
      local input = lookupTable:forward(sentences[i])
      local output = rnn:forward(input)
      local loss = criterion:forward(output, targets[i])
      local gradOutput = criterion:backward(output, targets[i])
      local gradInput = rnn:backward(input, gradOutput)
      return loss, gradParams
    end
    print('\nGradient checking, num params: ' .. params:size(1))
    local err = optim.checkgrad(feval, params)
    if err > 0.0001 then
      error('Gradient check failed. Error: ' .. err)  
    else
      print('Gradient check ok')
    end
  end
end
```

With this code, gradient check is failing, usually after 2-3 iterations of training. Here's the example output:

```
 [.................... 1/50 ....................]  ETA: 0ms | Step: 0ms         
Gradient checking, num params: 244	
Gradient check ok	
 [>................... 2/50 ....................]  ETA: 8m41s | Step: 10s858ms  
Gradient checking, num params: 244	
Gradient check ok	
 [=>.................. 3/50 ....................]  ETA: 6m24s | Step: 8s190ms   
Gradient checking, num params: 244	
/home/maciek/torch/install/bin/luajit: example-rnn.lua:51: Gradient check failed. Error: 0.19101544567184
stack traceback:
	[C]: in function 'error'
	example-rnn.lua:51: in main chunk
	[C]: in function 'dofile'
	...ciek/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00406670
```

However, when I change line 17 from:
`local seqLen = torch.random(maxSeqLen)`
to
`local seqLen = torch.random(maxSeqLen, maxSeqLen)`

basically forcing sequences to have same length(maxSeqLen), gradient check is no longer failing.
Is it a bug in rnn package or am I missing something here?

**Comment by nicholas-leonard** [comment]
Should be fixed via https://github.com/torch/nn/issues/759 once it is merged.

**Comment by nicholas-leonard** [comment]
@mpienkosz Thank you for the detailed bug report. Even though it isn't caused by rnn, I decided to include it as a unit test in https://github.com/Element-Research/rnn/commit/1e1a0489256a361454ba651f989a40019067e31a . Just in case :)

**Comment by mpienkosz** [comment]
@nicholas-leonard thank you for the response. I updated my rnn package but there still seems to be a problem. Your test seems to work as expected, but my code still fails on gradient check, this time even with sequences of the same length. 
I also tried to remove Sequencer module, replacing lines 25-30:
```
local rnn = nn.Sequential()
  :add(nn.SplitTable(1,2))
  :add(nn.Sequencer(nn.FastLSTM(inputSize, hiddenSize)))
  :add(nn.SelectTable(-1))
  :add(nn.Linear(hiddenSize, nClasses))
  :add(nn.LogSoftMax()) 
```
with:
```
local rnn = nn.Sequential()
  :add(nn.SplitTable(1,2))
  :add(nn.SelectTable(-1))
  :add(nn.Linear(inputSize, nClasses))
  :add(nn.LogSoftMax()) 
```
and then gradient check passed without errors, so I guess there is some issue with Sequencer. 
Could you please check it once again?

**Comment by nicholas-leonard** [comment]
@mpienkosz Did you update the nn package : `luarocks install nn`. The fix just got merged.

**Comment by nicholas-leonard** [comment]
@mpienkosz This should also help : https://github.com/torch/nn/pull/761 .

**Comment by mpienkosz** [comment]
@nicholas-leonard after this last SelectTable fix my code is no longer throwing any errors and works like a charm now;) Many thanks!

**Comment by nicholas-leonard** [comment]
@mpienkosz Yeah thanks for bringing this up!
